pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java

```diff
@@ -78,30 +87,53 @@ void validateAvro(Schema<?> schema) {
     }
   }
 
-  GenericRecord flattenGenericRecord(GenericRecord record) {
-    List<FieldValuePair> fieldValuePairs = buildFlattenedFields(record);
-    List<org.apache.avro.Schema.Field> fields =
-        fieldValuePairs.stream().map(FieldValuePair::getField).collect(Collectors.toList());
-    org.apache.avro.Schema modified = buildFlattenedSchema(record, fields);
-    GenericRecord newRecord = new GenericData.Record(modified);
-    fieldValuePairs.forEach(pair -> newRecord.put(pair.getField().name(), pair.getValue()));
+  private GenericRecord flattenGenericRecord(
+      GenericRecord record, Map<org.apache.avro.Schema, FlattenedSchema> schemaCache) {
+    FlattenedSchema modified =
+        schemaCache.computeIfAbsent(
+            record.getSchema(),
+            schema -> {
+              Map<org.apache.avro.Schema.Field, List<String>> flattenedFieldMap = new HashMap<>();
+              buildFlattenedFields(record.getSchema(), flattenedFieldMap);
+              List<org.apache.avro.Schema.Field> fields =
+                  new ArrayList<>(flattenedFieldMap.keySet());
+              return new FlattenedSchema(buildFlattenedSchema(record.getSchema(), fields), flattenedFieldMap);
+            });
+
+    GenericRecord newRecord = new GenericData.Record(modified.getSchema());
+    modified.getFieldMap().forEach((k, v) -> newRecord.put(k.name(), getFlattenedFieldValue(v, record)));
     return newRecord;
   }
 
-  private List<FieldValuePair> buildFlattenedFields(GenericRecord record) {
-    List<FieldValuePair> flattenedFields = new ArrayList<>();
-    org.apache.avro.Schema originalSchema = record.getSchema();
-    for (org.apache.avro.Schema.Field field : originalSchema.getFields()) {
-      flattenedFields.addAll(
-          flattenField(record, record.getSchema(), field, field.schema().isNullable(), ""));
+  private Object getFlattenedFieldValue(
```

Comment:
this can be further optimized, although maybe more complex than necessary.

Reply:
this improved ~10x with `parallelStream` on the loop calling this method - 118ms avg -> 17ms avg. 

```
    modified.getFieldMap().entrySet().parallelStream().forEach(e -> newRecord.put(e.getKey().name(), getFlattenedFieldValue(e.getValue(), record)));

```

after all, practical performance improvements (if needed) maybe a result of run some parts of code in parallel, rather than caching.